### PR TITLE
Handle parallel function calls for OpenAI LLMs

### PR DESCRIPTION
**Pull request by kwindla**
The hard work was done here by @jeevan057

  https://github.com/pipecat-ai/pipecat/pull/346#pullrequestreview-2336186908

I rebased, fixed one small bug, and modified the code from the original PR to use the newer context aggregator frames approach. I've tested this heavily.

There's one question in the comments of the `14-function-calling.py` example, but I don't think that should block a merge.

### CODE REVIEW

File: examples/foundational/14-function-calling.py [hunk 1]
```diff
@@ -106,11 +111,11 @@ async def main():
 
         pipeline = Pipeline(
             [
-                fl_in,
+                # fl_in,
                 transport.input(),
                 context_aggregator.user(),
                 llm,
-                fl_out,
+                # fl_out,
```
Comment (aconchillo) [inline examples/foundational/14-function-calling.py:118]: should we just remove the frame loggers?

File: examples/foundational/14-function-calling.py [hunk 2]
```diff
@@ -34,7 +34,12 @@
 
 
 async def start_fetch_weather(function_name, llm, context):
-    await llm.push_frame(TextFrame("Let me check on that."))
+    # note: we can't push a frame to the LLM here. the bot
+    # can interrupt itself and/or cause audio overlapping glitches.
+    # possible question for Aleix and Chad about what the right way
+    # to trigger speech is, now, with the new queues/async/sync refactors.
+    # await llm.push_frame(TextFrame("Let me check on that."))
+    logger.debug(f"Starting fetch_weather_from_api with function_name: {function_name}")
```
Comment (aconchillo) [inline examples/foundational/14-function-calling.py:42]: The rule is... if a processor can push frames from a different task then it should be async. But function calls happen in the same task, no? If so, this should be OK.

Comment (aconchillo) [reply examples/foundational/14-function-calling.py:42]: I just realized that, by default, all processors should be async and not the other way around. It makes more sense and it's safer this way. That was my initial idea but then changed my mine. But I think that was the right idea. I'll make the change in the morning and I believe that line should be safe.

File: examples/foundational/14-function-calling.py (hunk 1 above)
Comment (chadbailey59) [reply examples/foundational/14-function-calling.py:118]: Probaby. I want to leave them in an example or two so people know they exist, though.

File: examples/foundational/14-function-calling.py (hunk 2 above)
Comment (kwindla) [reply examples/foundational/14-function-calling.py:42]: I will rebase and test again after this change.

File: examples/foundational/14-function-calling.py (hunk 1 above)
Comment (aconchillo) [reply examples/foundational/14-function-calling.py:118]: If we leave them we should uncomment them I think.

**Review by chadbailey59 — APPROVED** [review]
I still want to do some downstream testing with RTVI, but I think this is good to merge 👍

**Pull request merged**